app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.spec.js

```diff
@@ -0,0 +1,134 @@
+'use strict';
+
+describe('providerSelectionService: API', () => {
+
+  let cloudProvider;
+  beforeEach(
+    window.module(
+      require('../cloudProvider.registry'),
+      function (cloudProviderRegistryProvider) {
+        cloudProvider = cloudProviderRegistryProvider;
+      }
+    )
+  );
+
+  beforeEach(window.module(require('./providerSelection.service'),
+    require('../../account/account.service')));
+
+  // required to ensure registry provider is available
+  let $q, $scope, $modal, accountService, cloudProviderRegistry, providerService;
+  beforeEach(
+    window.inject(
+      (_$q_, $rootScope, _$uibModal_, _accountService_, _cloudProviderRegistry_, _providerSelectionService_) => {
+        $q = _$q_;
+        $scope = $rootScope.$new();
+        $modal = _$uibModal_;
+        accountService = _accountService_;
+        cloudProviderRegistry = _cloudProviderRegistry_;
+        providerService = _providerSelectionService_;
+      }));
+
+  let hasValue, providers;
+  beforeEach(() => {
+    spyOn(accountService, 'listProviders').and.callFake(() => $q.when(providers));
+    spyOn(cloudProviderRegistry, 'hasValue').and.callFake(() => hasValue);
+    spyOn($modal, 'open').and.callFake(() => {
+      return {
+        result: $q.when('modalProvider')
+      };
+    });
+  });
+
+  beforeEach(() => {
+    window.spinnakerSettings.providers.testProvider = {
+      defaults: {
+        account: 'testProviderAccount',
+        region: 'testProviderRegion'
+      }
+    };
+  });
+
+  let application, config;
+  beforeEach(() => {
+
+    hasValue = false;
+    providers = [];
+    delete window.spinnakerSettings.defaultProvider;
+
+    application = {
+      name: 'testApplication',
+      attributes: {
+        cloudProviders: 'testProvider'
+      }
+    };
+
+    config = {
+      name: 'testProvider',
+      securityGroup: {}
+    };
+  });
+
+  it('should use the specified, default provider if the requested provider cannot be found', () => {
+    window.spinnakerSettings.defaultProvider = 'defaultProvider';
+
+    cloudProvider.registerProvider('fakeProvider', config);
+    providerService.selectProvider(application, 'securityGroup').then((_provider) => {
+      expect(_provider).toBe('defaultProvider');
```

Comment:
When testing async calls like this, I prefer to set a local variable, then move the `expect` after the `$scope.$digest()`. Otherwise, you can sometimes end up getting false positives if the promise is rejected or something else goes wrong in the underlying code.